sphinx/testing/_internal/markers.py

```diff
@@ -0,0 +1,364 @@
+    :param default_builder: The application default builder name.
+    :param default_isolation: The isolation default policy.
+    :param shared_result: An optional shared result ID.
+    :return: The application positional and keyword arguments.
+    """
+    # 1. process pytest.mark.sphinx
+    env = _get_sphinx_environ(node, default_builder)
+    # 1.1a. deduce the isolation policy from freshenv if possible
+    freshenv: bool | None = env.pop('freshenv', None)
+    if freshenv is not None:
+        if 'isolate' in env:
+            err = '%r and %r are mutually exclusive' % ('freshenv', 'isolate')
+            pytest.fail(format_mark_failure('sphinx', err))
+
+        isolation = env['isolate'] = Isolation.always if freshenv else default_isolation
+    else:
+        freshenv = env['freshenv'] = False
+
+    # 1.1b. deduce the final isolation policy
+    isolation = env.setdefault('isolate', default_isolation)
+    isolation = env['isolate'] = normalize_isolation_policy(isolation)
+    # 1.2. deduce the testroot ID
+    testroot_id = env['testroot'] = env.get('testroot', testroot_finder.default)
+    # 1.3. deduce the srcdir name (possibly explicitly given)
+    srcdir_name = env.get('srcdir', None)
+    srcdir = _get_test_srcdir(srcdir_name, testroot_id, shared_result)
+
+    # 2. process the srcdir ID according to the isolation policy
+    is_unique_srcdir_id = srcdir_name is not None
+    if isolation is Isolation.always:
+        # srcdir = XYZ-(RANDOM-UID)
+        srcdir = make_unique_id(srcdir)
+        is_unique_srcdir_id = True
+    elif isolation is Isolation.grouped:
+        if (location := get_node_location(node)) is None:
+            srcdir = make_unique_id(srcdir)
+            is_unique_srcdir_id = True
+        else:
+            # For a 'grouped' isolation, we want the same prefix (the deduced
+            # sources dierctory), but with a unique suffix based on the node
+            # location. In particular, parmetrized tests will have the same
+            # final ``srcdir`` value as they have the same location.
+            suffix = get_location_id(location)
+            # srcdir = XYZ-(RANDOM-UID)
+            srcdir = f'{srcdir}-{suffix}'
+
+    if is_unique_srcdir_id:
+        # when the sources directory is known to be unique across
+        # all other tests, we do not include a namespace or checksum
+        namespace, checksum = '-', 0
+    else:
+        namespace = get_container_id(node)
+        # Do a somewhat hash on configuration values to give a minimal protection
+        # against side-effects (two tests with the same configuration should have
+        # the same output; if they mess up with their sources directory, then they
+        # should be isolated accordingly). If there is a bug in the test suite, we
+        # can reduce the number of tests that can have dependencies by adding some
+        # isolation safeguards.
+        checksum = _get_common_config_checksum(env['buildername'], **env)
```

Comment:
Since checksumming selected elements from the Sphinx `env` should be relatively cheap, I think we could do this for unique `srcdir` values too.

However: perhaps it would be better to merge this feature separately before the plugin-related parts of the changeset?

Reply:
> I think we could do this for unique srcdir values too.

No because otherwise you can have very long paths. It doesn't really make sense to make something unique even more unique IMO (there is still a call to SHA-1 behind).